src/main/java/org/truffleruby/core/range/RubyRange.java

```diff
@@ -9,17 +9,36 @@
  */
 package org.truffleruby.core.range;
 
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
 import com.oracle.truffle.api.object.Shape;
 import org.truffleruby.core.klass.RubyClass;
 import org.truffleruby.language.RubyDynamicObject;
+import org.truffleruby.language.library.RubyLibrary;
 
+@ExportLibrary(RubyLibrary.class)
 public abstract class RubyRange extends RubyDynamicObject {
 
     public boolean excludedEnd;
+    public boolean frozen;
 
     public RubyRange(RubyClass rubyClass, Shape shape, boolean excludedEnd) {
+        this(rubyClass, shape, excludedEnd, true);
+    }
```

Comment:
Could you remove this constructor?
That way we'll make sure we'll have looked at all callers and reviewed them (e.g., `new RubyIntRange` in `.dup` is incorrect and could be easy to miss due to this).
Same for RubyIntRange and RubyLongRange, since we need to create some of them unfrozen, let's keep a single constructor and pass `frozen` explicitly everywhere when creating a new Range.